hack/lib.sh

```diff
@@ -0,0 +1,163 @@
+    --role roles/cloudscheduler.admin
+  gcloud projects add-iam-policy-binding ${project_id} \
+    --member=serviceAccount:${control_plane_service_account}@${project_id}.iam.gserviceaccount.com \
+    --role roles/logging.configWriter
+  gcloud projects add-iam-policy-binding ${project_id} \
+    --member=serviceAccount:${control_plane_service_account}@${project_id}.iam.gserviceaccount.com \
+    --role roles/logging.privateLogViewer
+
+}
+
+function init_pubsub_service_account() {
+  local project_id=${1}
+  local pubsub_service_account=${2}
+  echo "parameter project_id used when initiating data plane service account is'${project_id}'"
+  echo "parameter control_plane_service_account used when initiating pubsub service account is'${pubsub_service_account}'"
+  # Enable APIs.
+  gcloud services enable pubsub.googleapis.com
+
+  # Create the pubsub service account for the data plane if it doesn't exist
+  existing_pubsub_service_account=$(gcloud iam service-accounts list \
+    --filter="${SERVICE_ACCOUNT_EMAIL_KEY} ~ ^${pubsub_service_account}@")
+  if [ -z "${existing_pubsub_service_account}" ]; then
+    echo "Create PubSub Service Account '${pubsub_service_account}' neeeded for the Data Plane"
+    gcloud iam service-accounts create ${pubsub_service_account}
+  else
+    echo "PubSub Service Account needed for the Data Plane '${pubsub_service_account}' already existed"
+  fi
+
+  # Grant pubsub.editor role to the service account for the data plane to read and/or write to Pub/Sub.
+  gcloud projects add-iam-policy-binding ${project_id} \
+    --member=serviceAccount:${pubsub_service_account}@${project_id}.iam.gserviceaccount.com \
+    --role roles/pubsub.editor
+
+}
+
+function enable_workload_identity(){
+  local project_id=${1}
+  local control_plane_service_account=${2}
+  local cluster_location_type=${3}
+
+
+  echo "parameter project_id used when enabling workload identity is'${project_id}'"
+  echo "parameter control_plane_service_account used when when enabling workload identity is'${control_plane_service_account}'"
+  local cluster_location_option
+  if [ ${cluster_location_type} == ${ZONAL_CLUSTER_LOCATION_TYPE} ]; then
+    cluster_location_option=zone
+  elif [ ${cluster_location_type} == ${REGIONAL_CLUSTER_LOCATION_TYPE} ];then
+    cluster_location_option=region
+  else
+    echo >&2 "Fatal error: cluster_location_type used when when enabling workload identity has to '${ZONAL_CLUSTER_LOCATION_TYPE}' or '${REGIONAL_CLUSTER_LOCATION_TYPE}'"
+    exit 1
+  fi
+  echo "parameter cluster_location_type used when when enabling workload identity is'${cluster_location_type}'"
+  echo "cluster_location_option used when when enabling workload identity is'${cluster_location_option}'"
+  # Enable API
+  gcloud services enable iamcredentials.googleapis.com
+  # Enable workload identity.
+  echo "Enable Workload Identity"
+  local cluster_name="$(cut -d'_' -f4 <<<"$(kubectl config current-context)")"
```

Comment:
hmm, I don't think we can reliably parse the context to get cluster name and location. I believe user can change the context name.

Reply:
We need to add some comments like `The current project set in gcloud MUST be the same as where the cluster is running`

Reply:
Or do you think we should ask user for the project_id cluster name, cluster_location?

Reply:
I don't think we should ask the user. The script should be non-interactive.

We can use env vars for cluster name and location here, and only get values from kubectl context if those env vars are empty. I also recommend printing the chosen name and location so the user can see if the guess was correct.

Reply:
What does your `env vars` mean? Like this one **https://cloud.google.com/run/docs/configuring/environment-variables#command-line? Could you help provide an exmaple?

Reply:
Do you mean something like this 
`export E2E_PROJECT_ID="$(gcloud config get-value project)`
`export CLUSTER_NAME="$(gcloud config get-value run/cluster)`
`export CLUSTER_LOCATION="$(gcloud config get-value run/cluster_location)`

Reply:
Looks good. You can also use bash default values to allow the user to override:

```sh
PROJECT_ID=${PROJECT_ID:-$(gcloud config get-value project)}
CLUSTER_NAME=${CLUSTER_NAME:-$(gcloud config get-value run/cluster)}
CLUSTER_LOCATION=${CLUSTER_LOCATION:-$(gcloud config get-value run/cluster_location)}
```
```